frontend/catalyst/from_plxpr/decompose.py

```diff
@@ -0,0 +1,218 @@
+from copy import copy
+from typing import get_type_hints
+
+import jax
+import pennylane as qml
+
+# GraphSolutionInterpreter:
+from pennylane.decomposition import DecompositionGraph
+from pennylane.measurements import MidMeasureMP
+from pennylane.wires import WiresLike
+
+from catalyst.jax_primitives import decomposition_rule
+
+COMPILER_OPERATIONS_NUM_WIRES = {
+    "CNOT": 2,
+    "ControlledPhaseShift": 2,
+    "CRot": 2,
+    "CRX": 2,
+    "CRY": 2,
+    "CRZ": 2,
+    "CSWAP": 3,
+    "CY": 2,
+    "CZ": 2,
+    "Hadamard": 1,
+    "Identity": 1,
+    "IsingXX": 2,
+    "IsingXY": 2,
+    "IsingYY": 2,
+    "IsingZZ": 2,
+    "SingleExcitation": 2,
+    "DoubleExcitation": 4,
+    "ISWAP": 2,
+    "PauliX": 1,
+    "PauliY": 1,
+    "PauliZ": 1,
+    "PhaseShift": 1,
+    "PSWAP": 2,
+    "Rot": 1,
+    "RX": 1,
+    "RY": 1,
+    "RZ": 1,
+    "S": 1,
+    "SWAP": 2,
+    "T": 1,
+    "Toffoli": 3,
+    "U1": 1,
+    "U2": 1,
+    "U3": 1,
+}
+
+
+def create_decomposition_rule(func: Callable, op_name: str, num_wires: int):
+    """Create a decomposition rule from a function."""
+
+    sig_func = inspect.signature(func)
+    type_hints = get_type_hints(func)
+
+    args = {}
+    for name in sig_func.parameters.keys():
+        typ = type_hints.get(name, None)
+
+        # Skip tailing kwargs in the rules
+        if name == "__":
+            continue
+
+        if typ is float or name in ("phi", "theta", "omega", "delta"):
+            args[name] = float
+        elif typ is int:
+            args[name] = int
+        elif typ is WiresLike or name == "wires":
+            args[name] = qml.math.array([0] * num_wires, like="jax")
+        else:
+            raise ValueError(
+                f"Unsupported type annotation {typ} for parameter {name} in func {func}."
+            )
+
+    # Update the name of decomposition rule
+    rule_name = "_rule" if func.__name__[0] == "_" else "_rule_"
+    func.__name__ = op_name + rule_name + func.__name__ + "_wires_" + str(num_wires)
+
+    return decomposition_rule(func)(**args)
+
+
+# pylint: disable=too-few-public-methods
+class GraphSolutionInterpreter(qml.capture.PlxprInterpreter):
+    """Interpreter for getting the decomposition graph solution
+    from a jaxpr when program capture is enabled.
+    """
+
+    def __init__(
+        self,
+        *,
+        gate_set=None,
+        fixed_decomps=None,
+        alt_decomps=None,
+    ):  # pylint: disable=too-many-arguments
+
+        if not qml.decomposition.enabled_graph():
+            raise TypeError(
+                "The GraphSolutionInterpreter can only be used when"
+                "graph-based decomposition is enabled."
+            )
+
+        self._gate_set = gate_set
+        self._fixed_decomps = fixed_decomps
+        self._alt_decomps = alt_decomps
+
+        self._captured = False
+        self._operations = set()
+        self._decomp_graph_solution = {}
+
+    def interpret_operation(self, op: "qml.operation.Operator"):
```

Comment:
Wait, now that I look more closely, this `GraphSolutionInterpreter` doesn't actually decompose anything, its only purpose is to construct, solve a graph, and pre-compile all the decomposition rules? Have you thought about re-using the `CollectResourceOps` interpreter?

Reply:
It's not working with PLxPR and Catalyst... haha! 

This `GraphSolutionInterpreter` is designed to be used in `from_plxpr` interpreters